render/container.go

```diff
@@ -66,80 +73,67 @@ func ShortLivedConnectionJoin(r Renderer, toIPs func(report.Node) []string) Rend
-		id, ok := n.Latest.Lookup(originalNodeID)
-		if !ok {
-			return report.Nodes{}
-		}
-		topology, ok := n.Latest.Lookup(originalNodeTopology)
-		if !ok {
-			return report.Nodes{}
-		}
+	// Propagate the internet and service pseudo nodes
+	if strings.HasSuffix(n.ID, TheInternetID) || strings.HasPrefix(n.ID, ServiceNodeIDPrefix) {
+		return report.Nodes{n.ID: n}
+	}
 
-		return report.Nodes{
-			id: NewDerivedNode(id, n).
-				WithTopology(topology),
-		}
+	// If this node is not of the original type, exclude it.  This
+	// excludes all the nodes we've dragged in from endpoint that we
+	// failed to join to a node.
+	id, ok := n.Latest.Lookup(originalNodeID)
+	if !ok {
+		return report.Nodes{}
+	}
+	topology, ok := n.Latest.Lookup(originalNodeTopology)
+	if !ok {
+		return report.Nodes{}
 	}
 
-	// MapEndpoint2IP maps endpoint nodes to their IP address, for joining
-	// with container nodes.  We drop endpoint nodes with pids, as they
-	// will be joined to containers through the process topology, and we
-	// don't want to double count edges.
-	endpoint2IP := func(m report.Node, local report.Networks) report.Nodes {
-		// Don't include procspied connections, to prevent double counting
-		_, ok := m.Latest.Lookup(endpoint.Procspied)
-		if ok {
-			return report.Nodes{}
-		}
-		scope, addr, port, ok := report.ParseEndpointNodeID(m.ID)
-		if !ok {
-			return report.Nodes{}
-		}
+	return report.Nodes{
+		id: NewDerivedNode(id, n).
+			WithTopology(topology),
+	}
+}
 
-		if externalNode, ok := NewDerivedExternalNode(m, addr, local); ok {
-			return report.Nodes{externalNode.ID: externalNode}
-		}
+// endpoint2IP maps endpoint nodes to their IP address, for joining
+// with container nodes.
+func endpoint2IP(m report.Node, local report.Networks) report.Nodes {
+	scope, addr, port, ok := report.ParseEndpointNodeID(m.ID)
+	if !ok {
+		return report.Nodes{}
+	}
 
-		// We also allow for joining on ip:port pairs.  This is useful
-		// for connections to the host IPs which have been port
-		// mapped to a container can only be unambiguously identified with the port.
-		// So we need to emit two nodes, for two different cases.
-		id := report.MakeScopedEndpointNodeID(scope, addr, "")
-		idWithPort := report.MakeScopedEndpointNodeID(scope, addr, port)
-		return report.Nodes{
-			id:         NewDerivedNode(id, m).WithTopology(IP),
-			idWithPort: NewDerivedNode(idWithPort, m).WithTopology(IP),
-		}
+	if externalNode, ok := NewDerivedExternalNode(m, addr, local); ok {
+		return report.Nodes{externalNode.ID: externalNode}
 	}
 
-	return FilterUnconnected(MakeMap(
-		ipToNode,
-		MakeReduce(
-			MakeMap(
-				nodeToIP,
-				r,
-			),
-			MakeMap(
-				endpoint2IP,
-				SelectEndpoint,
-			),
-		),
-	))
+	// We also allow for joining on ip:port pairs.  This is useful for
+	// connections to the host IPs which have been port mapped to a
+	// container can only be unambiguously identified with the port.
+	// So we need to emit two nodes, for two different cases.
```

Comment:
Maybe mention that we need to do this because in some cases connection tracking doesn't provide PIDs?